src/requests/textdocument.jl

```diff
@@ -59,10 +52,20 @@ function process(r::JSONRPC.Request{Val{Symbol("textDocument/didClose")},DidClos
     doc = getdocument(server, URI2(uri))
     empty!(doc.diagnostics)
     publish_diagnostics(doc, server)
-    if !is_workspace_file(doc)
-        deletedocument!(server, URI2(uri))
-    else
+    if is_workspace_file(doc)
         set_open_in_editor(doc, false)
+    else
+        if any(d.root == doc.root && doc._open_in_editor for (uri, d::Document) in getdocuments_pair(server) if d != doc)
+            # If any other open document shares doc's root we just mark it as closed...
+            set_open_in_editor(doc, false)
+        else
+            # ...otherwise we delete all documents that share root with doc.
+            for (u,d) in getdocuments_pair(server)
+                if d.root == doc.root
```

Comment:
```suggestion
                if d.root == doc.root && !get_open_in_editor(d)
```
We need to make sure that we don't delete files here that are outside the workspace but are currently also open in the editor, I believe? If you accept this change here, you also need to add a `set_open_in_editor(doc, false)` before this loop, otherwise the doc that is actually being close won't get removed.

In general this seems all very tricky and hard to follow logic... if we do this here in `didClose`, do we need to do a similar thing in delete messages from file watch? I think this would all be way, way simpler if we went with https://github.com/julia-vscode/LanguageServer.jl/issues/544 :) I at least have a really hard time figuring out all the possible scenarios/cases here...